vsphere/resource_vsphere_tag_category.go

```diff
@@ -0,0 +1,251 @@
+		return fmt.Errorf("could not set associable type data for category: %s", err)
+	}
+
+	if err := d.Set("used_by", category.UsedBy); err != nil {
+		return fmt.Errorf("could not set user access info for category: %s", err)
+	}
+	return nil
+}
+
+func resourceVSphereTagCategoryUpdate(d *schema.ResourceData, meta interface{}) error {
+	client, err := meta.(*VSphereClient).TagsClient()
+	if err != nil {
+		return err
+	}
+
+	// Block the update if the user has removed types
+	oldts, newts := d.GetChange("associable_types")
+	for _, v1 := range oldts.(*schema.Set).List() {
+		var found bool
+		for _, v2 := range newts.(*schema.Set).List() {
+			if v1 == v2 {
+				found = true
+			}
+		}
+		if !found {
+			return fmt.Errorf("cannot remove type %q (removal of associable types is not supported)", v1)
+		}
+	}
+
+	id := d.Id()
+	spec := &tags.CategoryUpdateSpec{
+		UpdateSpec: tags.CategoryUpdate{
+			AssociableTypes: sliceInterfacesToStrings(d.Get("associable_types").(*schema.Set).List()),
+			Cardinality:     d.Get("cardinality").(string),
+			Description:     d.Get("description").(string),
+			Name:            d.Get("name").(string),
+		},
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), defaultAPITimeout)
+	defer cancel()
+	err = client.UpdateCategory(ctx, id, spec)
+	if err != nil {
+		return fmt.Errorf("could not update category with id %q: %s", id, err)
+	}
+	return resourceVSphereTagCategoryRead(d, meta)
+}
+
+func resourceVSphereTagCategoryDelete(d *schema.ResourceData, meta interface{}) error {
+	client, err := meta.(*VSphereClient).TagsClient()
+	if err != nil {
+		return err
+	}
+
+	id := d.Id()
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultAPITimeout)
+	defer cancel()
+	err = client.DeleteCategory(ctx, id)
+	if err != nil {
+		return fmt.Errorf("could not delete category with id %q: %s", id, err)
+	}
+	return nil
+}
+
+func resourceVSphereTagCategoryImport(d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+	// Although GetCategoriesByName does not seem to think that tag categories
+	// are unique, empirical observation via the console and API show that they
+	// are. Hence we can just import by the tag name. If for some reason the
+	// returned results includes more than one ID, we give an error.
+	client, err := meta.(*VSphereClient).TagsClient()
+	if err != nil {
+		return nil, err
+	}
+
+	name := d.Id()
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultAPITimeout)
+	defer cancel()
+	cats, err := client.GetCategoriesByName(ctx, name)
+	if err != nil {
+		return nil, fmt.Errorf("could not get category for name %q: %s", name, err)
+	}
+
+	if len(cats) < 1 {
+		return nil, fmt.Errorf("category name %q not found", name)
+	}
+	if len(cats) > 1 {
+		// This should not happen but guarding just in case it does.
+		return nil, fmt.Errorf("category name %q returned multiple results, more specific name required", name)
+	}
```

Comment:
why not combine these to be: 
```
if len(cats) == 1) {
  d.SetId(cats[0].ID)
 return []*schema.ResourceData{d}, nil 
}

return nil, fmt.Errorf("Expected a single category, got '%d'", len(cats))

Reply:
There's 2 specific cases here - the first one is when an incorrect category is given and the one that will be the one most users will encounter (`len(cats) < 1`). Technically, it seems impossible for `len(cats) > 1` to happen at the moment, as category names need to be unique (adding a category with a different name but different associable types in TF or the UI will result in a name collision error). Yet `GetCategoriesByName` [returns multiple](https://github.com/vmware/vic/blob/master/pkg/vsphere/tags/categories.go#L169), even though its own logic searches for an [exact match](https://github.com/vmware/vic/blob/master/pkg/vsphere/tags/categories.go#L183). So the second case is just for brevity to match the SDK and I don't expect it to really happen. Regardless, given these two cases are so distinct I'd prefer to have this level of verbosity over a more general `len != 1` style error message, which I think is more confusing...

Also as I'm typing this I think the second error is a little misleading too - might end up re-wording this as well, but we still need to distinguish between "not found" (something a user may stumble on if they just mis-typed the name of the category) versus `len > 1` (which is more of a change of the expected API and something we will need to fix).